Sources/MongoSwift/ReadPreference.swift

```diff
@@ -0,0 +1,164 @@
+        let maxStalenessSeconds = mongoc_read_prefs_get_max_staleness_seconds(self._readPreference)
+
+        return maxStalenessSeconds == MONGOC_NO_MAX_STALENESS ? nil : maxStalenessSeconds
+    }
+
+    /**
+     * Initializes a `ReadPreference` from a `Mode`.
+     *
+     * - Parameters:
+     *   - mode: a `Mode`
+     *
+     * - Returns: a new `ReadPreference`
+     */
+    public init(_ mode: Mode) {
+        self._readPreference = mongoc_read_prefs_new(mode.readMode)
+    }
+
+    /**
+     * Initializes a `ReadPreference`.
+     *
+     * - Parameters:
+     *   - mode: a `Mode`
+     *   - tagSets: an optional `[Document]`
+     *   - maxStalenessSeconds: an optional Int64
+     *
+     * - Returns: a new `ReadPreference`
+     *
+     * - Throws:
+     *   - A `MongoError.readPreferenceError` if mode is primary and tagSets is non-empty
+     *   - A `MongoError.readPreferenceError` if maxStalenessSeconds non-nil and < 90
+     */
+    public init(_ mode: Mode, tagSets: [Document]? = nil, maxStalenessSeconds: Int64? = nil) throws {
+        self._readPreference = mongoc_read_prefs_new(mode.readMode)
+
+        if let tagSets = tagSets {
+            guard mode != .primary || tagSets.isEmpty else {
+                throw MongoError.readPreferenceError(message: "tagSets may not be used with primary mode")
+            }
+
+            let encoder = BsonEncoder()
+            let tags = try encoder.encode(Document(tagSets))
+            mongoc_read_prefs_set_tags(self._readPreference, tags.data)
+        }
+
+        if let maxStalenessSeconds = maxStalenessSeconds {
+            guard maxStalenessSeconds >= MONGOC_SMALLEST_MAX_STALENESS_SECONDS else {
+                throw MongoError.readPreferenceError(message: "Expected maxStalenessSeconds to be >= \(MONGOC_SMALLEST_MAX_STALENESS_SECONDS), \(maxStalenessSeconds) given")
+            }
+
+            mongoc_read_prefs_set_max_staleness_seconds(self._readPreference, maxStalenessSeconds);
+        }
+    }
+
+    /// Initializes a new `ReadPreference` by copying an existing `ReadPreference`.
+    public init(from readPreference: ReadPreference) {
+        self._readPreference = mongoc_read_prefs_copy(readPreference._readPreference)
+    }
+
+    /// Initializes a new `ReadPreference` by copying a `mongoc_read_prefs_t`.
+    /// The caller is responsible for freeing the original `mongoc_read_prefs_t`.
+    internal init(from readPreference: OpaquePointer?) {
+        self._readPreference = mongoc_read_prefs_copy(readPreference)
+    }
+
+    /// Cleans up the internal `mongoc_read_prefs_t`.
+    deinit {
+        guard let readPreference = self._readPreference else { return }
+        mongoc_read_prefs_destroy(readPreference)
+        self._readPreference = nil
+    }
+}
+
+/// An extension of `ReadPreference` to make it `Equatable`.
+extension ReadPreference: Equatable {
+    public static func == (lhs: ReadPreference, rhs: ReadPreference) -> Bool {
```

Comment:
I have a slight stylistic preference to just write these like
```swift
return lhs.mode == rhs.mode
    && lhs.tagSets == rhs.tagSets
    && lhs.maxStalenessSeconds == rhs.maxStalenessSeconds
```

was there some reason for writing it like this instead?

Reply:
I just wasn't sure what the syntax was preferred for multi-line conditionals (haven't setup `swift-lint` locally) and I'm used to enumerating simple guard statements in PHPLIB. I'll change this to one condition.